addons/pkg-maven/common/src/main/java/org/commonjava/indy/pkg/maven/content/MetadataCacheManager.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (C) 2011-2018 Red Hat, Inc. (https://github.com/Commonjava/indy)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.pkg.maven.content;
+
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.pkg.maven.content.cache.MavenMetadataCache;
+import org.commonjava.indy.pkg.maven.content.cache.MavenMetadataKeyCache;
+import org.commonjava.indy.subsys.infinispan.CacheHandle;
+import org.infinispan.query.Search;
+import org.infinispan.query.dsl.Query;
+import org.infinispan.query.dsl.QueryFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@ApplicationScoped
+public class MetadataCacheManager
+{
+    protected final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    @MavenMetadataCache
+    private CacheHandle<MetadataKey, MetadataInfo> metadataCache;
+
+    @Inject
+    @MavenMetadataKeyCache
+    private CacheHandle<MetadataKey, MetadataKey> metadataKeyCache;
+
+    private QueryFactory queryFactory;
+
+    public MetadataCacheManager()
+    {
+    }
+
+    @PostConstruct
+    private void init()
+    {
+        this.queryFactory = Search.getQueryFactory( metadataKeyCache.getCache() );
+    }
+
+    public MetadataCacheManager( CacheHandle<MetadataKey, MetadataInfo> metadataCache,
+                                 CacheHandle<MetadataKey, MetadataKey> metadataKeyCache )
+    {
+        this.metadataCache = metadataCache;
+        this.metadataKeyCache = metadataKeyCache;
+        this.queryFactory = Search.getQueryFactory( metadataKeyCache.getCache() );
+    }
+
+    public void put( MetadataKey metadataKey, MetadataInfo metadataInfo )
```

Comment:
Should this method be synchronized?

Reply:
I think we don't need to. Whoever thread putting the value does not matter, we just use the latest one.

Reply:
> I think we don't need to. Whoever thread putting the value does not matter, we just use the latest one.

The race condition is not just in this method. It happens between remove and put. As now there are two operations for this metadata cache, If we don't control, one thread may remove a metakey just after a put operation from another one, which will cause chaos in the other operation. That's the race condition I'm concerning.

Reply:
Yeah, the argument is fair. If we sync the two methods, it would avoid some race between put and removal so the two caches are in sync. However, the ISPN expiration will bring new inconsistencies, if it expires a key from key cache but not from meta cache, and vice verse. 
Here comes my logic. Those two caches may not need to be 100% in sync. I.e., we can allow some keys not being in the meta cache, or vice verse. The key cache is mainly for quick removal of all paths for a certain storeKey. If it does leave some meta not cleared due to the inconsistency, that normally won't bring much trouble. After some time, the expiration will remove them anyway. 
We can fix the inconsistency completely by making the key cache the only source of truth, i.e., to check the key cache first when we get from meta cache. I haven't decided it yet. Adding a check may use a bit more time which a performance-sensitive coder want to avoid. But considering the metadata retrieval rate and ISPN being quick at getting, it might not be a problem. Anyway, I am swing between 100% accuracy and 99% inaccuracy with a bit of performance gain. I might toss a coin to help me. Anyway, thanks for thinking deeply into it and let me know if any new idea.